src/language/providers/closeUtils.ts

```diff
@@ -0,0 +1,495 @@
+        }
+      }
+    }
+  }
+  return false
+}
+
+export function checkMissingCloseTag(
+  document: vscode.TextDocument,
+  position: vscode.Position,
+  nsPrefix: string
+) {
+  const triggerLine = position.line
+  const triggerPos = position.character
+  const triggerText = document.lineAt(triggerLine).text
+  const itemsOnLine = getItemsOnLineCount(triggerText)
+
+  const items = [
+    'annotation',
+    'schema',
+    'element',
+    'sequence',
+    'choice',
+    'group',
+    'simpleType',
+    'complexType',
+    'assert',
+    'discriminator',
+    'defineVariable',
+    'setVariable',
+  ]
+
+  for (let i = 0; i < items.length; ++i) {
+    const textBeforeTrigger = triggerText.substring(0, triggerPos)
+
+    if (itemsOnLine > 1) {
+      if (textBeforeTrigger.lastIndexOf('<' + nsPrefix + items[i]) > -1) {
+        let gt1res = getItemsForLineGT1(
+          triggerText,
+          triggerPos,
+          nsPrefix,
+          items,
+          i
+        )
+
+        if (gt1res != undefined) {
+          return gt1res
+        }
+      }
+    }
+
+    if (itemsOnLine < 2) {
+      let lt2res = getItemsForLineLT2(
+        document,
+        triggerText,
+        triggerLine,
+        nsPrefix,
+        items,
+        i
+      )
+
+      if (lt2res != undefined) {
+        return lt2res
+      }
+    }
+  }
+
+  return 'none'
+}
+
+export function getCloseTag(
+  document: vscode.TextDocument,
+  position: vscode.Position,
+  nsPrefix: string,
+  tag: string,
+  startLine: number,
+  startPos: number
+): [string, number, number] {
+  let lineNum = startLine
+  let tagOpen = startPos
+  const triggerText = document.lineAt(startLine).text
+  const itemsOnLine = getItemsOnLineCount(document.lineAt(lineNum).text)
+  let tagPos = triggerText.indexOf('<')
+  let endPos = triggerText.lastIndexOf('>')
+
+  if (itemsOnLine > 1 && startPos !== tagPos && startPos < endPos) {
+    while (tagOpen > -1 && tagOpen < triggerText.length) {
+      tagOpen = triggerText.indexOf('<', tagOpen)
+      let tagClose = triggerText.indexOf('>', tagOpen)
+      let tagPart = triggerText.substring(tagOpen, tagClose)
+
+      if (tagPart.includes(tag) && tagPart.includes('/')) {
+        return [tag, startLine, tagOpen]
+      }
+
+      tagOpen = tagClose + 1
+    }
+  } else {
+    let endPos = triggerText.indexOf('>', startPos)
+
+    if (
+      (triggerText.includes('</') || triggerText.includes('/>')) &&
+      triggerText.includes(tag) &&
+      endPos > -1 &&
+      itemsOnLine < 2
+    ) {
+      return [tag, startLine, startPos]
+    }
+
+    ++lineNum
+
+    while (lineNum > -1 && lineNum < document.lineCount) {
+      let currentText = document.lineAt(lineNum).text
+
+      if (getItemsOnLineCount(currentText) < 2) {
+        if (currentText.includes('</' + nsPrefix + tag)) {
+          return [tag, lineNum, startPos]
+        }
+      }
+
+      ++lineNum
+    }
+  }
+
+  return ['none', 0, 0]
+}
+
+export function getItemsForLineGT1(
```

Comment:
@rt320 In reality this should probably be split up more but it was hard for to follow some of the looping going so it would be best that possibly you handle this since you know it better